docs/structured.md

```diff
@@ -0,0 +1,212 @@
+  for instance, will be in O(1) time. Adding a new index will take O(n) time,
+where n is the number of rows in the table.
+
+- Deletion: in O(n) where n is the number of subordinate entities (tables for a
+  namespace, rows for a table).
+
+The client API will be changed to use tabular based addressing instead of
+global-key addressing. Non-primary-keyed queries will be supported through the use of secondary indexes.
+
+##Non-Goals##
+
+For now we don't intend on developing the following:
+
+- Add support for triggers, stored-procedures, and integrity constraints
+
+- Add the notion of column types beyond *a sequence of bytes*. Add support for
+  unbounded sub-collections of columns.
+
+- Add support for non-tabular storage. This might be supported in the future.
+
+- Add support for column locality-groups to improve data layout for performance.
+
+##Design##
+
+We support high level entities like Namespace, Table, Index, and Column.
+{Namespace,Table}Descriptor stores the entity metadata for {Namespace,Table}.
+
+```proto
+NamespaceDescriptor {
+  NamespaceID = … ,
+  Permissions = …,
+  ...}
+```
+
+```proto
+TableDescriptor {
+  TableID = …,
+  Columns = [{ ColumnID = …, Name = ... }, ...],
+  Indexes = [{ IndexID = …, Name = …, ColumnIDs = [ … ]}, ...],
+  Permissions = …
+  NextFreeColumnId = …
+  NextFreeIndexID = ...,
+  ...}
+```
+
+In order to support fast renaming, we indirect entity access through a
+hierarchical name-to-identifier mapping. NamespaceID and TableID are
+allocated from the same global ID space. ColumnID and IndexID are local
+to each Table.
+
+To simplify our implementation, our Tables will require the presence of a
+primary key. Follow-up work may relax this requirement. An investigation of use
+cases not requiring a primary key is required to specify this work.
+
+Initially, all the Table metadata in a cluster will be distributed by gossip and
+cached by each cockroach node. To reduce contention on the Table metadata, a
+database transaction on the data within a Table will only contain the data being
+addressed, and will not include a database read of the Table metadata. Care will
+be taken to not expose these implementation details to the user.
+
+##Addressing: Anatomy of a key##
+
+The “/” separators used below are shorthand for ordered encoding of the
+separated values.
+
+###Global metadata key###
+
+The root namespace is an unnamed namespace with a fixed ID of 0. Within it,
+metadata addressing will work as follows:
+
+Namespace: `\x00ns<NamespaceName>` -> NamespaceDescriptor { NamespaceID = …, ...}
+
+Table: `\x00tbl<NamespaceID><TableName>` -> TableDescriptor { TableID = …,
+local-IDs, … }
+
+Creating/Renaming/Deleting: Creating a new Namespace or Table involves
+allocating a new ID, initializing the {Namespace,Table}Descriptor, and storing
+the descriptor at key:/parent-id/name. Renaming involves deleting the
+`/old-parent-id/old-name` and creating `/new-parent-id/new-name`. A namespace can
```

Comment:
The reason for my comment was that a namespace doesn't have a parent, and so parent-key would've more "flexible" and you could imagine it being "empty" if no parent existed. Not true for parentID, but I think people will figure out what's meant.